### ISSUE TITLE
Visual Debugger Support

### ISSUE THREAD

**Issue by githubpsyche** [quarto-dev/quarto#38]
QMD files currently don't have visual debugging support in the way Python scripts and Jupyter notebooks do. Without this feature, it's still much easier in VSCode to develop analyses using those other options.

In Interactive mode for Python scripts, You can click in the margin to the left of line numbers to set breakpoints. Then you can use Debug Cell to start a debugging session for that code cell. The debugger stops execution at breakpoints and allows you to step through code one line at a time and inspect variables (see [Debugging](https://code.visualstudio.com/docs/editor/debugging) for details.

The closest thing to this that I seem able to achieve is to add the line `import pdb; pdb.set_trace()` where I want to debug. This triggers a breakpoint where I can interrupt code flow to input commands, but I don't get to benefit from key affordances in VSCode's debugging UI.

**Comment by itcarroll** [comment]
I'm sure this would be part of the design, but it's crucial to be able to run in debug without dropping a breakpoint in the .qmd file itself; a breakpoint within imported code would still be picked up.

As an intermediate step, would it be possible to improve the experience with `%debug` magic? Currently, if running a cell produces an error, you can execute `%debug` in the interactive notebook ... but this drops you into ipdb using the VSCode Command Palette which is awkward. Would be better to end up in the debug console. Maybe that's a general VS Code issue? Happy to raise issue elsewhere if suggested.

**Comment by euklid321** [comment]
I think this feature would be a massive gamechanger for the VScode Quarto user base. The lack of it currently also prevents me from using Quarto markdowns in a lot of cases. Is there any chance of giving this some higher priority?

**Comment by githubpsyche** [comment]
I've been thinking about this lately. VSCode supports remote debugging if `.vscode/launch.json` is configured right. It might be possible to get most of VSCode's debugging features inside a QMD file if these details are figured out. You might not be able to set breakpoints the same way (i.e., you might have to add a line to your code that explicitly sets a debugging trace) but you might get access to the debugging console. If the quarto extension configures these settings automatically, that might be a sufficient stopgap until VSCode improves the interoperation of IPython with debugging. 

I'll spin up a demo!

**Comment by githubpsyche** [comment]
Yep, VSCode describes its best supported approach here (Python-specific) using `debugpy`:

https://code.visualstudio.com/docs/python/debugging#_local-script-debugging

```python
import debugpy

# 5678 is the default attach port in the VS Code debug configurations. Unless a host and port are specified, host defaults to 127.0.0.1
debugpy.listen(5678)
print("Waiting for debugger attach")
debugpy.wait_for_client()
debugpy.breakpoint()
print('break on this line')
```

I found that I don't even need to customize my `launch.json` for this to work.

The `import debugpy; debugpy.listen(5678); debugpy.wait_for_client()` stuff only runs once and sets up the connection with VSCode's debugger interface. From there, you can place `debugpy.breakpoint()` anywhere in your notebook to start a debugging session at that part of your code. You can even re-use `debugpy.breakpoint()` at multiple parts of your notebook and indeed seem to get access to variables created in other cells.

**Comment by githubpsyche** [comment]
Okay, to finalize, here's a minimal example adding a breakpoint to `hello.qmd`. The first code cell sets up the debugging session. Once you do, you can set breakpoints with `breakpoint()` inside your QMD file, or using VSCode's breakpoint UI inside imported python scripts. 

    ---
    title: "Quarto Basics"
    format:
      html:
        code-fold: true
    jupyter: python3
    ---

    For a demonstration of a line plot on a polar axis, see @fig-polar.

    ```{python}
    #| code-summary: code to start the debugger; attach to it in VS Code "Run and Debug" panel
    import debugpy
    debugpy.listen(5678)
    debugpy.wait_for_client()
    ```

    ```{python}
    #| label: fig-polar
    #| fig-cap: "A line plot on a polar axis"

    import numpy as np
    import matplotlib.pyplot as plt

    r = np.arange(0, 2, 0.01)
    theta = 2 * np.pi * r
    breakpoint() # minimum needed to connect to VSCode's debugger UI at this point
    fig, ax = plt.subplots(
      subplot_kw = {'projection': 'polar'} 
    )
    ax.plot(theta, r)
    ax.set_rticks([0.5, 1, 1.5, 2])
    ax.grid(True)
    plt.show()
    ```

I guess it's shame that VSCode doesn't already set up the debugging connection for you when you run something in its IPython panel.
